python/ray/util/spark/utils.py

```diff
@@ -412,17 +416,77 @@ def get_spark_task_assigned_physical_gpus(gpu_addr_list):
 
 
-def setup_sigterm_on_parent_death():
-    """
-    Uses prctl to automatically send SIGTERM to the child process when its parent is
-    dead. The child process itself should handle SIGTERM properly.
-    """
+# store the session folder path of current ray cluster.
+GLOBAL_RAY_CLUSTER_SESSION_NAME_FILE = "global_ray_cluster_session_name.txt"
+
+
+def _try_clean_temp_dir_at_exit(
+    process, collect_log_to_path, temp_dir, lock_fd=None, is_head_node=True
+):
+    if lock_fd is None:
+        lock_file = temp_dir + ".lock"
+        lock_fd = os.open(lock_file, os.O_RDWR | os.O_CREAT | os.O_TRUNC)
+
     try:
-        import ctypes
-        import signal
-
-        libc = ctypes.CDLL("libc.so.6")
-        # Set the parent process death signal of the command process to SIGTERM.
-        libc.prctl(1, signal.SIGTERM)  # PR_SET_PDEATHSIG, see prctl.h
-    except OSError as e:
-        _logger.warning(f"Setup libc.prctl PR_SET_PDEATHSIG failed, error {repr(e)}.")
+        # Wait for a while to ensure the children processes of the ray node all
+        # exited.
+        time.sleep(SIGTERM_GRACE_PERIOD_SECONDS + 0.5)
+        if process.poll() is None:
+            # "ray start ..." command process is still alive. Force to kill it.
+            process.kill()
+
+        # Release the shared lock, representing current ray node does not use the
+        # temp dir.
+        fcntl.flock(lock_fd, fcntl.LOCK_UN)
+
+        try:
+            # acquiring exclusive lock to ensure copy logs and removing dir safely.
+            fcntl.flock(lock_fd, fcntl.LOCK_EX | fcntl.LOCK_NB)
+            lock_acquired = True
+        except BlockingIOError:
+            # The file has active shared lock or exclusive lock, representing there
+            # are other ray nodes running, or other node running cleanup temp-dir
+            # routine. skip cleaning temp-dir, and skip copy logs to destination
+            # directory as well.
+            lock_acquired = False
+
+        if lock_acquired:
+            # This is the final terminated ray node on current spark worker,
+            # start copy logs (including all local ray nodes logs) to destination.
+            if collect_log_to_path:
+                try:
+                    copy_log_dest_path = os.path.join(
+                        collect_log_to_path,
+                        os.path.basename(temp_dir) + "-logs",
+                        socket.gethostname(),
+                    )
+                    if is_head_node:
+                        with open(
+                            os.path.join(
+                                temp_dir, GLOBAL_RAY_CLUSTER_SESSION_NAME_FILE
+                            ),
+                            "r",
+                        ) as f:
+                            ray_session_dir = f.read()
+                    else:
+                        ray_session_dir = os.readlink(
+                            os.path.join(temp_dir, "session_latest")
+                        )
```

Comment:
We need to wait until node is started then 'session_latest' is ready so that we can get the session directory. On head node, https://github.com/ray-project/ray/blob/57dfd4325bab5b29e60815f98d6cc599a66ecbd3/python/ray/util/spark/cluster_init.py#L139-L143 this makes sure the head node is already started and I write down the session directory into a local file. On workers, I feel like normally it should just have one ray session, as if users run `ray start` it only create ray cluster on driver not on workers, unless they explicitly write some tricky code to start ray clusters on worker, I think this code should be fine. @WeichenXu123